instrumentation-genai/opentelemetry-instrumentation-langchain/src/opentelemetry/instrumentation/langchain/callback_handler.py

```diff
@@ -0,0 +1,240 @@
+import time
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional
+from uuid import UUID
+
+from langchain_core.callbacks import BaseCallbackHandler  # type: ignore
+from langchain_core.messages import BaseMessage  # type: ignore
+from langchain_core.outputs import LLMResult  # type: ignore
+
+from opentelemetry.context import Context, get_current
+from opentelemetry.instrumentation.langchain.utils import dont_throw
+from opentelemetry.semconv._incubating.attributes import (
+    gen_ai_attributes as GenAI,
+)
+from opentelemetry.semconv.attributes import (
+    error_attributes as ErrorAttributes,
+)
+from opentelemetry.trace import Span, SpanKind, Tracer, set_span_in_context
+from opentelemetry.trace.status import Status, StatusCode
+
+
+@dataclass
+class _SpanState:
+    span: Span
+    span_context: Context
+    start_time: float = field(default_factory=time.time)
+    children: List[UUID] = field(default_factory=list)
+
+
+class OpenTelemetryLangChainCallbackHandler(BaseCallbackHandler):  # type: ignore[misc]
+    """
+    A callback handler for LangChain that uses OpenTelemetry to create spans for LLM calls and chains, tools etc,. in future.
+    """
+
+    def __init__(
+        self,
+        tracer: Tracer,
+    ) -> None:
+        super().__init__()  # type: ignore
+        self._tracer = tracer
+
+        # Map from run_id -> _SpanState, to keep track of spans and parent/child relationships
+        self.spans: Dict[UUID, _SpanState] = {}
+        self.run_inline = True  # Whether to run the callback inline.
+
+    def _create_span(
```

Comment:
Maybe a style nit, but another option would be to have this handler class only implement the interface methods (single responsibility). These other methods could go into a separate class (e.g. a SpanManger or something)